File: searchlib/src/vespa/searchlib/aggregation/quantile_aggregation_result.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
#include "quantile_aggregation_result.h"

#include <vespa/searchlib/expression/resultvector.h>
#include <vespa/vespalib/objects/serializer.hpp>
#include <vespa/vespalib/objects/deserializer.hpp>

#include <vespa/log/log.h>
LOG_SETUP(".searchlib.aggregation.quantile_aggregation_result");

namespace search::aggregation {

using expression::ResultNodeVector;
using vespalib::Serializer;
using vespalib::Deserializer;

IMPLEMENT_IDENTIFIABLE_NS2(search, aggregation, QuantileAggregationResult, AggregationResult);

QuantileAggregationResult::QuantileAggregationResult() {
    _rank.reset(new FloatResultNode(0));
}

QuantileAggregationResult::~QuantileAggregationResult() = default;

QuantileAggregationResult::QuantileAggregationResult(const QuantileAggregationResult&) = default;

QuantileAggregationResult& QuantileAggregationResult::operator=(const QuantileAggregationResult&) = default;

void QuantileAggregationResult::visitMembers(vespalib::ObjectVisitor& visitor) const {
    AggregationResult::visitMembers(visitor);
    visit(visitor, "quantiles", _quantiles);

    for (const auto [quantile, value] : quantile_results()) {
        visit(visitor, std::format("quantile({})", quantile), value);
    }

    visit(visitor, "extension", _extension);
}

void QuantileAggregationResult::onPrepare(const ResultNode& result, bool useForInit) {
    (void) result;
    if (useForInit) {
        LOG(warning, "useForInit was true. Should not happen for QuantileAggregationResult.");
    }
}

void QuantileAggregationResult::onMerge(const AggregationResult& b) {
    _sketch.merge(dynamic_cast<const QuantileAggregationResult&>(b)._sketch);
}

void QuantileAggregationResult::onAggregate(const ResultNode& result) {
    if (result.isMultiValue()) {
        auto& rv = dynamic_cast<const ResultNodeVector &>(result);
        for (size_t i = 0; i < rv.size(); ++i) {
            _sketch.update(rv.get(i).getFloat());
        }
    } else {
        _sketch.update(result.getFloat());
    }
}

void QuantileAggregationResult::onReset() {
    _sketch = vespalib::KLLSketch();
}

Serializer& QuantileAggregationResult::onSerialize(vespalib::Serializer& os) const {
    AggregationResult::onSerialize(os);
    os << _quantiles;

    os << _extension;

    const std::vector<uint8_t> buf = _sketch.serialize();
    os << buf;

    return os;
}

Deserializer& QuantileAggregationResult::onDeserialize(Deserializer& is) {
    AggregationResult::onDeserialize(is);
    is >> _quantiles;

    is >> _extension;

    std::vector<uint8_t> buf;
    is >> buf;
    _sketch = vespalib::KLLSketch::deserialize(buf);

    return is;
}

std::vector<QuantileAggregationResult::QuantileResult> QuantileAggregationResult::quantile_results() const {
    std::vector<QuantileResult> results;
    results.reserve(_quantiles.size());

    for (const auto quantile : _quantiles) {
        double value = 0.0;
        if (!_sketch.is_empty()) {
            value = _sketch.get_quantile(quantile);
        }
        results.emplace_back(quantile,  value);
    }

    return results;
}

void QuantileAggregationResult::set_quantiles(const std::vector<double>& quantiles) {
    _quantiles = quantiles;
}

const std::vector<double>& QuantileAggregationResult::quantiles() const {
    return _quantiles;
}

void QuantileAggregationResult::update_sketch(double value) {
    _sketch.update(value);
}
} // namespace search::aggregation

File: searchlib/src/main/java/com/yahoo/searchlib/aggregation/QuantileAggregationResult.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.searchlib.aggregation;

import com.yahoo.searchlib.expression.FloatResultNode;
import com.yahoo.searchlib.expression.ResultNode;
import com.yahoo.vespa.objects.Deserializer;
import com.yahoo.vespa.objects.ObjectVisitor;
import com.yahoo.vespa.objects.Serializer;
import org.apache.datasketches.kll.KllDoublesSketch;
import org.apache.datasketches.memory.Memory;

import java.util.Arrays;

/**
 * This is an aggregated result holding the specified quantile.
 *
 * @author johsol
 */
public class QuantileAggregationResult extends AggregationResult {

    public static final int classId = registerClass(0x4000 + 179, QuantileAggregationResult.class, QuantileAggregationResult::new);

    private double quantile;
    private KllDoublesSketch sketch;

    /**
     * Constructs an empty result node. <b>NOTE:</b> This instance is broken until non-optional member data is set.
     */
    public QuantileAggregationResult() {

    }

    /**
     * Constructs an instance of this class with given quantile.
     */
    public QuantileAggregationResult(double quantile) {
        this.quantile = quantile;
        this.sketch = KllDoublesSketch.newHeapInstance();
    }

    public double getQuantile() {
        return quantile;
    }

    public double getValue() {
        if (sketch.isEmpty()) {
            return 0;
        }
        return sketch.getQuantile(quantile);
    }

    public QuantileAggregationResult setQuantile(double quantile) {
        this.quantile = quantile;
        return this;
    }

    public QuantileAggregationResult updateSketch(double value) {
        this.sketch.update(value);
        return this;
    }

    @Override
    public ResultNode getRank() {
        return new FloatResultNode(sketch.getQuantile(quantile));
    }

    @Override
    protected void onMerge(AggregationResult result) {
        sketch.merge(((QuantileAggregationResult) result).sketch);
    }

    @Override
    protected int onGetClassId() {
        return classId;
    }

    @Override
    protected void onSerialize(Serializer buf) {
        super.onSerialize(buf);
        buf.putDouble(null, quantile);
        byte[] bytes = this.sketch.toByteArray();

        buf.putInt(null, bytes.length);
        buf.put(null, bytes);
    }

    @Override
    protected void onDeserialize(Deserializer buf) {
        super.onDeserialize(buf);
        quantile = buf.getDouble(null);
        int length = buf.getInt(null);
        byte[] bytes = buf.getBytes(null, length);
        sketch = KllDoublesSketch.heapify(Memory.wrap(bytes));
    }

    @Override
    protected boolean equalsAggregation(AggregationResult obj) {
        return (quantile == ((QuantileAggregationResult) obj).quantile) && Arrays.equals(sketch.toByteArray(), ((QuantileAggregationResult) obj).sketch.toByteArray());
    }

    @Override
    public QuantileAggregationResult clone() {
        QuantileAggregationResult obj = (QuantileAggregationResult) super.clone();
        obj.quantile = quantile;
        obj.sketch = KllDoublesSketch.heapify(Memory.wrap(sketch.toByteArray()));
        return obj;
    }

    @Override
    public void visitMembers(ObjectVisitor visitor) {
        super.visitMembers(visitor);
        visitor.visit("quantile", quantile);
        visitor.visit("value", getValue());
    }
}
